src/workerd/jsg/jsg.h

```diff
@@ -1326,17 +1326,44 @@ using PromiseForResult = Promise<RemovePromise<ReturnType<Func, Param, passLock>
 
 class ModuleRegistry;
 
+class ContextGlobal {
+  // All jsg resources that are used as context globals must eventually extend this type.
+  // The lifecycle of this object is tied to a v8 context.
+
+public:
+  ContextGlobal() {}
+
+  KJ_DISALLOW_COPY_AND_MOVE(ContextGlobal);
+
+  ModuleRegistry& getModuleRegistry() { return *moduleRegistry; }
+
+private:
+  kj::Own<ModuleRegistry> moduleRegistry;
+
+  void setModuleRegistry(kj::Own<ModuleRegistry> registry) {
+    moduleRegistry = kj::mv(registry);
+  }
+
+  template <typename, typename>
+  friend class ResourceWrapper;
+};
+
+class ContextGlobalObject: public Object, public ContextGlobal { };
```

Comment:
Nit: this could just be a convenience macro rather than a separate declaration.

e.g.
```cpp
#define JSG_CONTEXT_GLOBAL public Object, public ContextGlobal

class Foo : JSG_CONTEXT_GLOBAL { } 
```

Or, to help with the `EventTarget` case... something like

```cpp
#define JSG_CONTEXT_GLOBAL_OBJECT(Base) public Base, public ContextGlobal

class Foo : JSG_CONTEXT_GLOBAL(jsg::Object) {} 
class Bar : JSG_CONTEXT_GLOBAL(EventTarget) {}
```